qiskit/algorithms/evolvers/pvqd/pvqd.py

```diff
@@ -0,0 +1,390 @@
+        evolution: Optional[EvolutionSynthesis] = None,
+        use_parameter_shift: bool = True,
+        quantum_instance: Optional[Union[Backend, QuantumInstance]] = None,
+    ) -> None:
+        """
+        Args:
+            ansatz: A parameterized circuit preparing the variational ansatz to model the
+                time evolved quantum state.
+            initial_parameters: The initial parameters for the ansatz. Together with the ansatz,
+                these define the initial state of the time evolution.
+            timestep: The time step.
+            optimizer: The classical optimizers used to minimize the overlap between
+                Trotterization and ansatz. Can be either a :class:`.Optimizer` or a callable
+                using the :class:`.Minimizer` protocol.
+            expectation: The expectation converter to evaluate expectation values.
+            initial_guess: The initial guess for the first VQE optimization. Afterwards the
+                previous iteration result is used as initial guess. If None, this is set to
+                a random vector with elements in the interval :math:`[-0.01, 0.01]`.
+            evolution: The evolution synthesis to use for the construction of the Trotter step.
+                Defaults to first-order Lie-Trotter decomposition, see also
+                :mod:`~qiskit.synthesis.evolution` for different options.
+            use_parameter_shift: If True, use the parameter shift rule to compute gradients.
+                If False, the optimizer will not be passed a gradient callable. In that case,
+                Qiskit optimizers will use a finite difference rule to approximate the gradients.
+            quantum_instance: The backend of quantum instance used to evaluate the circuits.
+        """
+        if evolution is None:
+            evolution = LieTrotter()
+
+        self.ansatz = ansatz
+        self.initial_parameters = initial_parameters
+        self.timestep = timestep
+        self.optimizer = optimizer
+        self.initial_guess = initial_guess
+        self.expectation = expectation
+        self.evolution = evolution
+        self.gradients = use_parameter_shift
+
+        self._sampler = None
+        self.quantum_instance = quantum_instance
+
+    @property
+    def quantum_instance(self) -> Optional[QuantumInstance]:
+        """Return the current quantum instance."""
+        return self._quantum_instance
+
+    @quantum_instance.setter
+    def quantum_instance(self, quantum_instance: Optional[Union[Backend, QuantumInstance]]) -> None:
+        """Set the quantum instance and circuit sampler."""
+        if quantum_instance is not None:
+            if not isinstance(quantum_instance, QuantumInstance):
+                quantum_instance = QuantumInstance(quantum_instance)
+            self._sampler = CircuitSampler(quantum_instance)
+
+        self._quantum_instance = quantum_instance
+
+    def step(
```

Comment:
A number of these methods are public, `step`, `get_loss` etc. Are we allowing  end users to call these such that they can do their own evolve like function any conditions or should these be private.

I guess its partly confusing as I see anstaz is passed in here to step which when coming from evolve is self.ansatz and the validate_setup ensures that has been set. If I call here and pass in ansatz then get_loss does a validate_setup and even though it gets given the anstaz just passed will complain if one has not been set on the object (ie. self.ansatz is still None) as far as I can see.

If these are wanted as public I think maybe we should have some unit tests that directly call these. Yes they are tested indirectly via evolve but the path is slightly different code as I note with ansatz above.

Reply:
Yes they're supposed to be public so users can disect the algorithm and do the steps themselves. I'll add some unittests 👍🏻